src/H5C.c

```diff
@@ -953,6 +982,352 @@ H5C_evict(H5F_t *f)
+ *              hash bucket in the page_index, and then scanning the
+ *              contents of the bucket for entries residing in the
+ *              target page.
+ *
+ *              For each such entry, we test to see if it is pinned.
+ *              If it is not, we simply evict it.
+ *
+ *              Pinned entries may in turn be divided into tagged and
+ *              un-tagged entries.
+ *
+ *              For pinned tagged entries, it would be best if we could
+ *              simply tell the associated cache client to refresh it.
+ *              However, until we have that facility, we look up its tag,
+ *              and evict all entries associated with that on disk object.
+ *
+ *              For pinned, un-tagged entries (i.e. super block, global
+ *              heaps, etc. we must instruct the client to refresh the
+ *              entry.  Fortunately, this is only necessary for the
+ *              super block in the initial VFD SWMR implementation.
+ *
+ *              Note that there is also the possibility that while the
+ *              page was modified, one or more metadata entries in
+ *              that page were not.  Eventually we should write code
+ *              to detect this -- but not for the prototype.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ * Programmer:  John Mainzer -- 12/16/18
+ *
+ * Changes:     Added macro calls to maintain the page buffer hints.
+ *
+ *                                           JRM -- 3/20/20
+ *
+ *-------------------------------------------------------------------------
+ */
+
+herr_t
+H5C_evict_or_refresh_all_entries_in_page(H5F_t *f, uint64_t page, uint32_t length, uint64_t tick)
+{
+    int                i;
+    size_t             image_len;
+    size_t             original_image_len;
+    void *             image_ptr     = NULL;
+    void *             new_image_ptr = NULL;
+    unsigned           flush_flags   = (H5C__FLUSH_INVALIDATE_FLAG | H5C__FLUSH_CLEAR_ONLY_FLAG);
+    haddr_t            tag;
+    H5C_t *            cache_ptr = NULL;
+    H5C_cache_entry_t *entry_ptr;
+    H5C_cache_entry_t *follow_ptr = NULL;
+    herr_t             ret_value  = SUCCEED; /* Return value */
+
+    FUNC_ENTER_NOAPI(FAIL)
+
+    /* Sanity check */
+    HDassert(f);
+    HDassert(f->shared);
+
+    cache_ptr = f->shared->cache;
+
+    HDassert(cache_ptr);
+    HDassert(cache_ptr->magic == H5C__H5C_T_MAGIC);
+    HDassert(cache_ptr->vfd_swmr_reader);
+
+    /* since file must be opened R/O for a VFD SWMR reader, the skip
+     * list must be empty.  Verify this.
+     */
+    HDassert(cache_ptr->slist_len == 0);
+
+    i = H5C__PI_HASH_FCN(page);
+
+    entry_ptr = (cache_ptr->page_index)[i];
+
+    while (entry_ptr) {
+
+        HDassert(entry_ptr->magic == H5C__H5C_CACHE_ENTRY_T_MAGIC);
+
+        if (entry_ptr->page == page) {
+
+            HDassert(entry_ptr->addr >= (haddr_t)(page * cache_ptr->page_size));
```

Comment:
The haddr_t casts on these multiplies should occur before the uint64_t is multiplied.   Someday when haddr_t is 128-bits, this code will not work correctly.